### DISCUSSION TITLE
Python <3.9 Syntax for Inheriting Generic Type

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [microsoft/pyright#3970] [Q&A] [answered]
I'm using Python 3.8, and I've seen:

- [PyLance reports subscript will generate runtime exception within TYPE_CHECKING clause](https://github.com/microsoft/pylance-release/issues/2728)
- [Type subscript runtime exception false positive](https://github.com/microsoft/pylance-release/issues/1145)

What is the syntax for defining a class that inherits from a Generic type?

I've tried the following:

```python
# Python 3.8
from __future__ import annotations

from collections import OrderedDict
from typing_extensions import TypeAlias

od_alias: TypeAlias = 'OrderedDict[str, int]'

class mydict(OrderedDict[str, int]):  # Subscript for class "OrderedDict" will generate runtime exception; enclose type annotation in quotes
    pass

class mydict(OrderedDict['str', 'int']):  # Subscript for class "OrderedDict" will generate runtime exception; enclose type annotation in quotes
    pass

class mydict(od_alias):  # Inheriting 'od_alias', which is not a class
    pass
```

**Comment by adam-grant-hendry** [comment, marked as answer]
See this answered here: [python discussion](https://discuss.python.org/t/python-3-9-syntax-for-inheriting-generic-type/19308/3)

```python
from __future__ import annotations
```

provides delayed evaluation of _annotations_ ([PEP 563](https://peps.python.org/pep-0563/) and [PEP 649](https://peps.python.org/pep-0649/)). Base classes and the right hand side (RHS) of a `TypeAlias` statement are not annotations.

A language change was required to get the index notation to work, which came about in Python 3.9. For Python <3.9, one should use

```python
from typing import OrderedDict
```

this applies also to the builtins, like `list`, `dict` and so on (specifically when used as a base class in a class definition).

**Comment by erictraut** [comment]
Here's the way to do this if you need to be compatible with Python 3.8.

```python
from typing import OrderedDict

class mydict(OrderedDict[str, int]):
    pass
```

**Comment by hmc-cs-mdrissi** [comment]
I frequently do this for classes that lack runtime generic support (untyped 3rd party libraries) and tend to rely on TYPE_CHECKING tricks.

```python
if TYPE_CHECKING:
  base_cls: TypeAlias = "Foo[int]"
else:
  base_cls = Foo

class MyFoo(base_cls):
  ...
```

Occasionally I need to do something even messier for case of subclass is generic on type that was generic in parent at stub time but not run time. That hack looks like,

```python
if TYPE_CHECKING:
  FooBase = Foo
else:
  class FooBase(Foo, Generic[T]):
    ...

class MyFoo(FooBase[T]):
  ...
```
